Migrating H2 to MySQL fails with "Column 'sizeX' cannot be null"

##- Ubuntu Bionic
- H2 and Cloud SQL MySQL 5.7
- Metabase version: 0.32.4
- Metabase hosting environment: Docker with a mounted database volume on Linux/Ubuntu Bionic (not running Docker during the migration)
- Metabase internal database: H2 (default)

- *Repeatable steps to reproduce the issue*
Unfortunately I cannot share the database that is producing the issue.
```
> java -jar metabase.jar load-from-h2 /path/to/metabase.db
log4j:ERROR Could not find value for key log4j.appender.metabase
log4j:ERROR Could not instantiate appender named "metabase".
04-10 18:17:07 INFO metabase.util :: Loading Metabase...
04-10 18:17:07 INFO metabase.util :: Maximum memory available to JVM: 3.7 GB
04-10 18:17:12 INFO util.encryption :: Saved credentials encryption is DISABLED for this Metabase instance. 🔓
 For more information, see https://metabase.com/docs/latest/operations-guide/start.html#encrypting-your-database-connection-details-at-rest
04-10 18:17:15 INFO metabase.driver :: Registered abstract driver :sql  🚚
04-10 18:17:19 INFO metabase.db :: Verifying mysql Database Connection ...
04-10 18:17:19 DEBUG plugins.classloader :: Setting current thread context classloader to NEWLY CREATED classloader clojure.lang.DynamicClassLoader@657d2501...
04-10 18:17:19 INFO metabase.driver :: Registered abstract driver :sql-jdbc (parents: :sql) 🚚
Load driver :sql-jdbc took 10 ms
04-10 18:17:19 INFO metabase.driver :: Registered driver :mysql (parents: :sql-jdbc) 🚚
Load driver :mysql took 79 ms
04-10 18:17:19 INFO metabase.driver :: Initializing driver :sql...
04-10 18:17:19 INFO metabase.driver :: Initializing driver :sql-jdbc...
04-10 18:17:19 INFO metabase.driver :: Initializing driver :mysql...
04-10 18:17:19 INFO metabase.db :: Verify Database Connection ...  ✅
04-10 18:17:19 INFO metabase.db :: Running Database Migrations...
04-10 18:17:19 INFO metabase.db :: Setting up Liquibase...
04-10 18:17:19 INFO metabase.db :: Liquibase is ready.
04-10 18:17:19 INFO metabase.db :: Checking if Database has unrun migrations...
04-10 18:17:21 INFO metabase.db :: Database Migrations Current ...  ✅
Temporarily disabling DB constraints...
[OK]
Transfering 9 instances of Database....[OK]
Transfering 29 instances of User....[OK]
Transfering 16 instances of Setting....[OK]
Transfering 8703 instances of Table.................................[OK]
Transfering 100839 instances of Field....................................................................................................................................................................................................................................................................................................................................................[OK]
Transfering 21870 instances of FieldValues............................................................................[OK]
Transfering 1 instances of Segment....[OK]
Transfering 392 instances of Revision.....[OK]
Transfering 1665 instances of ViewLog.........[OK]
Transfering 115 instances of Session....[OK]
Transfering 14 instances of Dashboard....[OK]
Transfering 52 instances of Card....[OK]
Transfering 1 instances of CardFavorite....[OK]
Transfering 47 instances of DashboardCard....BatchUpdateException:
 Message: (conn=40206) Column 'sizeX' cannot be null
 SQLState: 23000
 Error Code: 1048
java.sql.BatchUpdateException: (conn=40206) Column 'sizeX' cannot be null
	at org.mariadb.jdbc.MariaDbStatement.executeBatchExceptionEpilogue(MariaDbStatement.java:289)
	at org.mariadb.jdbc.MariaDbPreparedStatementClient.executeBatch(MariaDbPreparedStatementClient.java:306)
	at clojure.java.jdbc$execute_batch.invokeStatic(jdbc.clj:591)
	at clojure.java.jdbc$execute_batch.invoke(jdbc.clj:584)
	at clojure.java.jdbc$db_do_execute_prepared_statement$fn__12510.invoke(jdbc.clj:1039)
	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:842)
	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:769)
	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:782)
	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:769)
	at clojure.java.jdbc$db_do_execute_prepared_statement.invokeStatic(jdbc.clj:1038)
	at clojure.java.jdbc$db_do_execute_prepared_statement.invoke(jdbc.clj:1024)
	at clojure.java.jdbc$db_do_prepared.invokeStatic(jdbc.clj:1062)
	at clojure.java.jdbc$db_do_prepared.invoke(jdbc.clj:1042)
	at clojure.java.jdbc$insert_cols_BANG_.invokeStatic(jdbc.clj:1576)
	at clojure.java.jdbc$insert_cols_BANG_.invoke(jdbc.clj:1567)
	at clojure.java.jdbc$insert_multi_BANG_.invokeStatic(jdbc.clj:1633)
	at clojure.java.jdbc$insert_multi_BANG_.invoke(jdbc.clj:1601)
	at metabase.cmd.load_from_h2$insert_entity_BANG_$fn__47368.invoke(load_from_h2.clj:141)
	at metabase.cmd.load_from_h2$insert_entity_BANG_.invokeStatic(load_from_h2.clj:140)
	at metabase.cmd.load_from_h2$insert_entity_BANG_.invoke(load_from_h2.clj:123)
	at metabase.cmd.load_from_h2$load_data_BANG_.invokeStatic(load_from_h2.clj:154)
	at metabase.cmd.load_from_h2$load_data_BANG_.invoke(load_from_h2.clj:148)
	at metabase.cmd.load_from_h2$load_from_h2_BANG_$fn__47488.invoke(load_from_h2.clj:235)
	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:799)
	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:769)
	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:834)
	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:769)
	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:782)
	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:769)
	at metabase.cmd.load_from_h2$load_from_h2_BANG_.invokeStatic(load_from_h2.clj:228)
	at metabase.cmd.load_from_h2$load_from_h2_BANG_.invoke(load_from_h2.clj:220)
	at clojure.lang.Var.invoke(Var.java:384)
	at metabase.cmd$load_from_h2.invokeStatic(cmd.clj:37)
	at metabase.cmd$load_from_h2.invoke(cmd.clj:30)
	at clojure.lang.AFn.applyToHelper(AFn.java:154)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.core$apply.invokeStatic(core.clj:665)
	at clojure.core$apply.invoke(core.clj:660)
	at metabase.cmd$run_cmd$fn__53180.invoke(cmd.clj:140)
	at metabase.cmd$run_cmd.invokeStatic(cmd.clj:140)
	at metabase.cmd$run_cmd.invoke(cmd.clj:136)
	at clojure.lang.Var.invoke(Var.java:388)
	at metabase.core$run_cmd.invokeStatic(core.clj:128)
	at metabase.core$run_cmd.invoke(core.clj:126)
	at metabase.core$_main.invokeStatic(core.clj:137)
	at metabase.core$_main.doInvoke(core.clj:133)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at metabase.core.main(Unknown Source)
Caused by: java.sql.SQLIntegrityConstraintViolationException: (conn=40206) Column 'sizeX' cannot be null
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.get(ExceptionMapper.java:229)
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.getException(ExceptionMapper.java:165)
	at org.mariadb.jdbc.MariaDbStatement.executeBatchExceptionEpilogue(MariaDbStatement.java:286)
	... 47 more
Caused by: java.sql.SQLException: Column 'sizeX' cannot be null
Query is: INSERT INTO report_dashboardcard ( `sizeX`, `updated_at`, `col`, `id`, `parameter_mappings`, `card_id`, `visualization_settings`, `dashboard_id`, `created_at`, `sizeY`, `row` ) VALUES ( <null>, '2018-11-06 20:02:25.43', 0, 12, '[{"parameter_id":"4c972b75","card_id":13,"target":["dimension",["template-tag","project"]]},{"parameter_id":"55fb9bb0","card_id":13,"target":["dimension",["template-tag","created_at"]]}]', 13, '{}', 1, '2018-04-23 22:49:07.548', <null>, 5 )
java thread: MariaDb-bulk-4
	at org.mariadb.jdbc.internal.util.LogQueryTool.exceptionWithQuery(LogQueryTool.java:126)
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol$1.handleResultException(AbstractQueryProtocol.java:625)
	at org.mariadb.jdbc.internal.protocol.AsyncMultiRead.call(AsyncMultiRead.java:142)
	at org.mariadb.jdbc.internal.protocol.AsyncMultiRead.call(AsyncMultiRead.java:66)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:844)
Command failed with exception: (conn=40206) Column 'sizeX' cannot be null
```

--- a/src/metabase/cmd/load_from_h2.clj
+++ b/src/metabase/cmd/load_from_h2.clj
@@ -14,14 +14,11 @@
    mysql -u root -e 'DROP DATABASE IF EXISTS metabase; CREATE DATABASE metabase;'
    MB_DB_TYPE=mysql MB_DB_HOST=localhost MB_DB_PORT=3305 MB_DB_USER=root MB_DB_DBNAME=metabase lein run load-from-h2
    ```"
-  (:require [clojure
-             [set :as set]
-             [string :as str]]
-            [clojure.java
+  (:require [clojure.java
              [io :as io]
              [jdbc :as jdbc]]
+            [clojure.string :as str]
             [colorize.core :as color]
-            [medley.core :as m]
             [metabase
              [db :as mdb]
              [util :as u]]
@@ -58,6 +55,7 @@
              [table :refer [Table]]
              [user :refer [User]]
              [view-log :refer [ViewLog]]]
+            [metabase.util.i18n :refer [trs]]
             [toucan.db :as db])
   (:import java.sql.SQLException))
 
@@ -120,37 +118,50 @@
 
 ;;; ------------------------------------------- Fetching & Inserting Rows --------------------------------------------
 
-(defn- insert-entity! [target-db-conn {:keys [table], :as entity} objs]
-  ;; TODO - I don't think the print+flush is working as intended :/
-  (print (u/format-color 'blue "Transfering %d instances of %s..." (count objs) (:name entity)))
-  (flush)
-  ;; 1) `:sizeX` and `:sizeY` come out of H2 as `:sizex` and `:sizey` because of automatic lowercasing; fix the
-  ;;    names of these before putting into the new DB
+(defn- objects->colums+values
+  "Given a sequence of objects/rows fetched from the H2 DB, return a the `columns` that should be used in the `INSERT`
+  statement, and a sequence of rows (as seqeunces)."
+  [objs]
+  ;; 1) `:sizeX` and `:sizeY` come out of H2 as `:sizex` and `:sizey` because of automatic lowercasing; fix the names
+  ;;    of these before putting into the new DB
+  ;;
   ;; 2) Need to wrap the column names in quotes because Postgres automatically lowercases unquoted identifiers
-  (let [ks   (keys (set/rename-keys (first objs) {:sizex :sizeX, :sizey :sizeY}))
-        cols (for [k ks]
-               ((db/quote-fn) (name k)))]
-    ;; The connection closes prematurely on occasion when we're inserting thousands of rows at once. Break into
-    ;; smaller chunks so connection stays alive
-    (doseq [chunk (partition-all 300 objs)
-            :let  [vals (for [row chunk]
-                          (map row ks))]]
-      (print (color/blue \.))
-      (flush)
-      (try
-        (jdbc/insert-multi! target-db-conn table cols vals)
-        (catch SQLException e
-          (jdbc/print-sql-exception-chain e)
-          (throw e)))))
-  (println-ok))
+  (let [source-keys (keys (first objs))
+        dest-keys   (for [k source-keys]
+                      ((db/quote-fn) (name (case k
+                                             :sizex :sizeX
+                                             :sizey :sizeY
+                                             k))))]
+    {:cols dest-keys
+     :vals (for [row objs]
+             (map (comp u/jdbc-clob->str row) source-keys))}))
 
+(def ^:private chunk-size 100)
+
+(defn- insert-chunk! [target-db-conn table-name chunkk]
+  (print (color/blue \.))
+  (flush)
+  (try
+    (let [{:keys [cols vals]} (objects->colums+values chunkk)]
+      (jdbc/insert-multi! target-db-conn table-name cols vals))
+    (catch SQLException e
+      (jdbc/print-sql-exception-chain e)
+      (throw e))))
+
+(defn- insert-entity! [target-db-conn {table-name :table, entity-name :name} objs]
+  (print (u/format-color 'blue "Transfering %d instances of %s..." (count objs) entity-name))
+  (flush)
+  ;; The connection closes prematurely on occasion when we're inserting thousands of rows at once. Break into
+  ;; smaller chunks so connection stays alive
+  (doseq [chunk (partition-all chunk-size objs)]
+    (insert-chunk! target-db-conn table-name chunk))
+  (println-ok))
 
 (defn- load-data! [target-db-conn h2-connection-string-or-nil]
   (jdbc/with-db-connection [h2-conn (h2-details h2-connection-string-or-nil)]
-    (doseq [e     entities
-            :let  [rows (for [row (jdbc/query h2-conn [(str "SELECT * FROM " (name (:table e)))])]
-                          (m/map-vals u/jdbc-clob->str row))]
-            :when (seq rows)]
+    (doseq [{table-name :table, :as e} entities
+            :let                       [rows (jdbc/query h2-conn [(str "SELECT * FROM " (name table-name))])]
+            :when                      (seq rows)]
       (insert-entity! target-db-conn e rows))))
 
 
@@ -224,6 +235,9 @@
   Defaults to using `@metabase.db/db-file` as the connection string."
   [h2-connection-string-or-nil]
   (mdb/setup-db!)
+
+  (assert (#{:postgres :mysql} (mdb/db-type))
+    (str (trs "Metabase can only transfer data from H2 to Postgres or MySQL/MariaDB.")))
 
   (jdbc/with-db-transaction [target-db-conn (mdb/jdbc-details)]
     (jdbc/db-set-rollback-only! target-db-conn)

--- a/test/metabase/cmd/load_from_h2_test.clj
+++ b/test/metabase/cmd/load_from_h2_test.clj
@@ -1,8 +1,13 @@
 (ns metabase.cmd.load-from-h2-test
   (:require [expectations :refer [expect]]
+            [flatland.ordered.map :as ordered-map]
+            [metabase
+             [db :as mdb]
+             [util :as u]]
             [metabase.cmd.load-from-h2 :as load-from-h2]
-            [metabase.util :as u]
-            [toucan.models :as models]))
+            [toucan
+             [db :as db]
+             [models :as models]]))
 
 ;; Make sure load-from-h2 works with or without `file:` prefix
 (expect
@@ -48,3 +53,40 @@
 (expect
   (all-model-names)
   (migrated-model-names))
+
+;; make sure `objects->colums+values` properly handles the columns with weird casing: `sizeX` and `sizeY`
+(expect
+  {:cols ["\"id\"" "\"row\"" "\"sizeX\"" "\"sizeY\""]
+   :vals [[281 0 18 9]]}
+  (binding [db/*quoting-style* :ansi]
+    (-> (#'load-from-h2/objects->colums+values
+         ;; using ordered-map so the results will be in a predictable order
+         [(ordered-map/ordered-map
+           :id    281
+           :row   0
+           :sizex 18
+           :sizey 9)])
+        (update :cols vec))))
+
+;; make sure `objects->colums+values` properly de-CLOBs and CLOBs (by calling `u/jdbc-clob->str`)
+
+(defrecord ^:private FakeClob [s])
+
+(expect
+  {:cols ["\"created_at\"" "\"description\"" "\"parameter_mappings\"" "\"visualization_settings\""]
+   :vals [[#inst "2019-04-05T21:26:39.936-00:00"
+           "This is a description"
+           []
+           {}]]}
+  (binding [db/*quoting-style* :ansi]
+    (with-redefs [mdb/db-type      (constantly :postgres)
+                  u/jdbc-clob->str #(cond-> %
+                                      (instance? FakeClob %) :s)]
+      (-> (#'load-from-h2/objects->colums+values
+           [(ordered-map/ordered-map
+             :created_at             #inst "2019-04-05T21:26:39.936000000-00:00"
+             :description            (FakeClob. "This is a description")
+             :parameter_mappings     []
+             :visualization_settings {})])
+          (update :cols vec)
+          (update :vals vec)))))

